In [26]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import  classification_report,  ConfusionMatrixDisplay
from matplotlib.pylab import rcParams
#pip install liac-arff
import arff


# Exploratory data analysis
## Preprocessing
### Replacing strings in the arff

Especially this line is important:

`line.replace('	', '').replace(',,',',').replace('ckd,\n', 'ckd\n').replace(' yes','yes')`

otherwise, the file could not be opened

In [37]:
fin = open("data/chronic_kidney_disease_full.arff", "rt")
#output file to write the result to
fout = open("preprocessed/chronic_kidney_disease_full.arff", "wt")
#for each line in the input file
for line in fin:
	#read replace the string and write to output file
	fout.write(line.replace('	', '').replace(',,',',').replace('ckd,\n', 'ckd\n').replace(' yes','yes'))
#close input and output files
fin.close()

dataset = arff.load(open("preprocessed/chronic_kidney_disease_full.arff", "r"))
print(dataset['description'])

1. Title: Early stage of Indians Chronic Kidney Disease(CKD)

2. Source Information:
  (a) Source:
Dr.P.Soundarapandian.M.D.,D.M
    (Senior Consultant Nephrologist),
Apollo  Hospitals,
Managiri,
Madurai Main Road,
Karaikudi,
Tamilnadu,
India.
  (b) Creator:
L.Jerlin Rubini(Research Scholar)
Alagappa University
EmailId   :jel.jerlin@gmail.com
ContactNo :+91-9597231281
  (c) Guided by:
Dr.P.Eswaran Assistant Professor,
Department of Computer Science and Engineering,
Alagappa University,
Karaikudi,
Tamilnadu,
India.
Emailid:eswaranperumal@gmail.com
  (d) Date     : july 2015

3.Relevant Information:
age-age
bp-blood pressure
sg-specific gravity
al-   albumin
su-sugar
rbc-red blood cells
pc-pus cell
pcc-pus cell clumps
ba-bacteria
bgr-blood glucose random
bu-blood urea
sc-serum creatinine
sod-sodium
pot-potassium
hemo-hemoglobin
pcv-packed cell volume
wc-white blood cell count
rc-red blood cell count
htn-hypertension
dm-diabetes mellitus
cad-coronary artery disease
appet-appetite
pe-pedal

## Investigating the attributes

In [40]:
for attr in dataset['attributes']:
    print(attr[0], attr[1])

age NUMERIC
bp NUMERIC
sg ['1.005', '1.010', '1.015', '1.020', '1.025']
al ['0', '1', '2', '3', '4', '5']
su ['0', '1', '2', '3', '4', '5']
rbc ['normal', 'abnormal']
pc ['normal', 'abnormal']
pcc ['present', 'notpresent']
ba ['present', 'notpresent']
bgr NUMERIC
bu NUMERIC
sc NUMERIC
sod NUMERIC
pot NUMERIC
hemo NUMERIC
pcv NUMERIC
wbcc NUMERIC
rbcc NUMERIC
htn ['yes', 'no']
dm ['yes', 'no']
cad ['yes', 'no']
appet ['good', 'poor']
pe ['yes', 'no']
ane ['yes', 'no']
class ['ckd', 'notckd']


In [61]:
df = pd.DataFrame(dataset['data'])

df

0     1      2  3  4       5         6           7           8      9   \
0    48.0  80.0  1.020  1  0    None    normal  notpresent  notpresent  121.0   
1     7.0  50.0  1.020  4  0    None    normal  notpresent  notpresent    NaN   
2    62.0  80.0  1.010  2  3  normal    normal  notpresent  notpresent  423.0   
3    48.0  70.0  1.005  4  0  normal  abnormal     present  notpresent  117.0   
4    51.0  80.0  1.010  2  0  normal    normal  notpresent  notpresent  106.0   
..    ...   ...    ... .. ..     ...       ...         ...         ...    ...   
395  55.0  80.0  1.020  0  0  normal    normal  notpresent  notpresent  140.0   
396  42.0  70.0  1.025  0  0  normal    normal  notpresent  notpresent   75.0   
397  12.0  80.0  1.020  0  0  normal    normal  notpresent  notpresent  100.0   
398  17.0  60.0  1.025  0  0  normal    normal  notpresent  notpresent  114.0   
399  58.0  80.0  1.025  0  0  normal    normal  notpresent  notpresent  131.0   

     ...    15      16   17   18   19  20    21   22   23      24  
0    ...  44.0  7800.0  5.2  yes  yes  no  good   no   no     ckd  
1    ...  38.0  6000.0  NaN   no   no  no  good   no   no     ckd  
2    ...  31.0  7500.0  NaN   no  yes  no  poor   no  yes     ckd  
3    ...  32.0  6700.0  3.9  yes   no  no  poor  yes  yes     ckd  
4    ...  35.0  7300.0  4.6   no   no  no  good   no   no     ckd  
..   ...   ...     ...  ...  ...  ...  ..   ...  ...  ...     ...  
395  ...  47.0  6700.0  4.9   no   no  no  good   no   no  notckd  
396  ...  54.0  7800.0  6.2   no   no  no  good   no   no  notckd  
397  ...  49.0  6600.0  5.4   no   no  no  good   no   no  notckd  
398  ...  51.0  7200.0  5.9   no   no  no  good   no   no  notckd  
399  ...  53.0  6800.0  6.1   no   no  no  good   no   no  notckd  

[400 rows x 25 columns]

### None values in categorical attributes

In [52]:
for idx, attr in enumerate(dataset['attributes']):
    if attr[1] != 'NUMERIC':
        print(df.columns[idx], attr[0], df[df.columns[idx]].unique())

2 sg ['1.020' '1.010' '1.005' '1.015' None '1.025']
3 al ['1' '4' '2' '3' '0' None '5']
4 su ['0' '3' '4' '1' None '2' '5']
5 rbc [None 'normal' 'abnormal']
6 pc ['normal' 'abnormal' None]
7 pcc ['notpresent' 'present' None]
8 ba ['notpresent' 'present' None]
18 htn ['yes' 'no' None]
19 dm ['yes' 'no' None]
20 cad ['no' 'yes' None]
21 appet ['good' 'poor' None]
22 pe ['no' 'yes' None]
23 ane ['no' 'yes' None]
24 class ['ckd' 'notckd']


### Nan values in numerical attributes


In [60]:
for idx, attr in enumerate(dataset['attributes']):
    if attr[1] == 'NUMERIC':
        print(df.columns[idx], attr[0], np.count_nonzero(df[df.columns[idx]].isna().values), 'Nan values')

0 age 9 Nan values
1 bp 12 Nan values
9 bgr 44 Nan values
10 bu 19 Nan values
11 sc 17 Nan values
12 sod 87 Nan values
13 pot 88 Nan values
14 hemo 52 Nan values
15 pcv 71 Nan values
16 wbcc 106 Nan values
17 rbcc 131 Nan values


- wbcc has 106 nan values out of 400 in total (White blood cell count)
- rbcc has 131 nan values out of 400 in total (Red blood cell count)
  